### ISSUE TITLE
UltiSnips freezes after hitting <S-Del>in insert mode (neovim and vim)

### ISSUE THREAD

**Issue by KoviRobi** [SirVer/ultisnips#1307]
This is annoying as I often accidentally press S-Del instead of S-Ins (to insert clipboard contents into the buffer.

**Expected behavior:**
The text `<S-Del>` is inserted into the buffer.

**Actual behavior:**
Vim/Neovim freezes. Vim additionally (after freezing and pressing a key to get into insert mode and another key to insert, e.g. pressing "i" twice) outputs:
```
Error detected while processing InsertCharPre Autocommands for "*"..script /nix/store/mdwjzss9247y195r7winwnz21b1z9ias-vimplugin-ultisnips-2020-09-08/share/vim-plugins/ultisnips/autoload/UltiSnips.vim:
line    7:
E319: Sorry, the command is not available in this version: py3 import vim
Press ENTER or type command to continue
Error detected while processing InsertCharPre Autocommands for "*"..script /nix/store/mdwjzss9247y195r7winwnz21b1z9ias-vimplugin-ultisnips-2020-09-08/share/vim-plugins/ultisnips/autoload/UltiSnips.vim:
line    8:
E319: Sorry, the command is not available in this version: py3 from UltiSnips import UltiSnips_Manager
Press ENTER or type command to continue

















Error detected while processing InsertCharPre Autocommands for "*"..function UltiSnips#TrackChange:
line    1:
E319: Sorry, the command is not available in this version:     py3 UltiSnips_Manager._track_change()
Press ENTER or type command to continue
Error detected while processing TextChangedI Autocommands for "*"..function UltiSnips#TrackChange:
line    1:
E319: Sorry, the command is not available in this version:     py3 UltiSnips_Manager._track_change()
Press ENTER or type command to continue
```

For nvim, I have to kill the python process that nvim starts to get any response. Then it inserts `<S-Del>` into the buffer, and outputs:
```
Error detected while processing function UltiSnips#TrackChange[1]..provider#python3#Call:
line   18:
Error invoking 'python_execute' on channel 3 (python3-script-host):
ch 3 was closed by the client
Press ENTER or type command to continue
Error detected while processing function UltiSnips#TrackChange[1]..provider#python3#Call:
line   18:
Error invoking 'python_execute' on channel 3 (python3-script-host):
Invalid channel: 3
Error detected while processing function UltiSnips#TrackChange[1]..provider#python3#Call:
line   18:
Error invoking 'python_execute' on channel 3 (python3-script-host):
Invalid channel: 3
Press ENTER or type command to continue
Error detected while processing function UltiSnips#TrackChange[1]..provider#python3#Call:
line   18:
Error invoking 'python_execute' on channel 3 (python3-script-host):
Invalid channel: 3
Press ENTER or type command to continue
Error detected while processing function UltiSnips#TrackChange[1]..provider#python3#Call:
line   18:
Error invoking 'python_execute' on channel 3 (python3-script-host):
Invalid channel: 3
Press ENTER or type command to continue
Error detected while processing function UltiSnips#TrackChange[1]..provider#python3#Call:
line   18:
Error invoking 'python_execute' on channel 3 (python3-script-host):
Invalid channel: 3
Press ENTER or type command to continu
```

**Steps to reproduce**

1. Install vim/neovim
2. Add UltiSnips
3. Open vim/neovim
4. Go into insert mode
5. Hold shift, press delete, release both

Here is how I tested both
```nix
$ cat <<EOF > nvim-ultisnips-test.nix
{ pkgs ? import <nixpkgs> { } }:

with pkgs;
mkShell {
  buildInputs = [
    (neovim.override {
      configure = {
        packages.myVimPackage = with vimPlugins; {
          start = [ UltiSnips ];
          opt = [ ];
        };
      };
    })
  ];
}
EOF
$ nix-shell --pure nvim-ultisnips-test.nix
nix-shell $ nvim
# Press "i" then "<S-Del>"
```

Or for vim
```nix
$ cat <<EOF > vim-ultisnips-test.nix
{ pkgs ? import <nixpkgs> { } }:

with pkgs;
mkShell {
  buildInputs = [
    (vim_configurable.customize {
      name = "vim-ultisnips-test";
      vimrcConfig.packages.myVimPackage = with vimPlugins; {
        start = [ UltiSnips ];
        opt = [ ];
      };
    })
  ];
}
EOF
$ nix-shell --pure vim-ultisnips-test.nix
nix-shell $ vim-ultisnips-test
# Press "i" then "<S-Del>"
```

-----

- **Operating System**: NixOS 21.03.git.22dd0914a20 (Okapi)
- **Vim Version**:
   - nvim:
      ```
      NVIM v0.4.4
      Build type: Release
      LuaJIT 2.1.0-beta3
      ```
   - vim:
      ```
      VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Jan 01 1980 00:00:00)
      Included patches: 1-1522
      ```
- **UltiSnips Version**: 
   ```nix
   { version = "2020-09-08";
     src = fetchFromGitHub {
       owner = "SirVer";
       repo = "ultisnips";
       rev = "7941f98337564d4e86cf2b7467b64def4d212794";
       sha256 = "0miihbj9h3ay6iq1ld13ixij3l4h02yrmc7bq8iarw776l89gin0";
     };
   }
   ```
- **Python inside Vim**: 
   - vim: `2.7.18 (default, Apr 19 2020, 21:45:35) [GCC 9.3.0]` and `E319: Sorry, the command is not available in this version`
   - nvim: `2.7.18 (default, Apr 19 2020, 21:45:35) [GCC 9.3.0]` and `3.8.5 (default, Jul 20 2020, 13:26:22) [GCC 9.3.0]`
- **Docker repo/vimrc**: No docker, sorry, but there is a Nix expression, evaluated with nixpkgs `22dd0914a20`. If you want I can generate a docker container from the nix expressions though!

**Comment by mmngreco** [comment]
The same here! 

Killing this process I can get the control back with problems but enough to save my work and to restart neovim.

```bash
/home/.../envs/neovim/bin/python -c import sys; sys.path.remove(""); import neovim; neovim.start_host() script_host.py
```

**Comment by bertilow** [comment]
I get this too, but not with shitft-Del. I happens sometimes, but I can't pinpoint what triggers it. The only solution is to kill the offending python3 process, and restart Neovim. Very annoying.

**Comment by seb-jones** [comment]
I seem to get similar behaviour occasionally when changing tabs (e.g. using `:tabnext`, `:tabprevious` etc.) - nvim hangs for a while and then shows a bunch of errors like this:

```
Error detected while processing function UltiSnips#CursorMoved[1]..provider#python3#Call:
line   18:
Error invoking 'python_execute' on channel 5 (python3-script-host):
ch 5 was closed by the client
Error detected while processing function UltiSnips#LeavingBuffer[5]..provider#python3#Call:
line   18:
Error invoking 'python_execute' on channel 5 (python3-script-host):
Invalid channel: 5
```

It doesn't happen consistently, though, and I don't know if there is a simple way to replicate it. Seems to just occur at random.

**Comment by SirVer** [comment]
I could not reproduce this with nix-shell on my Mac. Can I has a docker please?

**Comment by KoviRobi** [comment]
That's because (assuming you are just using `import <nixpkgs>` which imports the latest nixpkgs) it seems to now have been fixed, I'm doing a quick bisect to see where the fix was introduced :)

**Comment by KoviRobi** [comment]
The commit that fixed this in nixpkgs is:

```diff
commit 965ac833ccefba31ebc6bb164f3b9821adf2c559
Author: R. RyanTM <ryantm-bot@ryantm.com>
Date:   Sat Oct 3 12:25:41 2020 +0000

    python27Packages.pynvim: 0.4.1 -> 0.4.2

diff --git a/pkgs/development/python-modules/pynvim/default.nix b/pkgs/development/python-modules/pynvim/default.nix
index 93cba65ae5e..53bb06b13dd 100644
--- a/pkgs/development/python-modules/pynvim/default.nix
+++ b/pkgs/development/python-modules/pynvim/default.nix
@@ -12,11 +12,11 @@
 
 buildPythonPackage rec {
   pname = "pynvim";
-  version = "0.4.1";
+  version = "0.4.2";
 
   src = fetchPypi {
     inherit pname version;
-    sha256 = "0n2cx22lrmbq7xk7356lyn6k77ryqvkxplw9k0fglk35ckb1isam";
+    sha256 = "6bc6204d465de5888a0c5e3e783fe01988b032e22ae87875912280bef0e40f8f";
   };
 
   nativeBuildInputs = [
```

which incidentally, has the changelog:

> Pynvim 0.4.2
> 
> Changes since 0.4.1:
> 
>     1276622 Refactor: IOError was deprecated. Replace with OSError
>     a5446ce refactor: Reduce usage of deprecated imp module
>     605ffda Modify all imports in the repository to absolute imports
>     d9ce0b5 Set unicode_errors in Unpacker
>     2cbc135 import vim module by default

so I guess it wasn't related to UltiSnips directly, but maybe UltiSnips just called a function that had this bug

If you want to play around with it, this nix-expression will drop you into the shell where nvim has this problem (I haven't disected the vim-problem, I suspect it might be a different commit):

```nix
{ bad-pkgs ? import
    (builtins.fetchTarball {
      name = "nixos-unstable-2020-09-28";
      url = "https://github.com/nixos/nixpkgs/archive/046c6a7038998507134981c8a294bb70861484bf.tar.gz";
      # Hash obtained using `nix-prefetch-url --unpack <url>`
      sha256 = "03qbkh3n6gd8lqy6gc0d8nxvxgg3w6i02x592q8f703vqcl20vd2";
    })
    { }
, good-pkgs ? import
    (builtins.fetchTarball {
      name = "nixos-unstable-2020-10-03";
      url = "https://github.com/nixos/nixpkgs/archive/965ac833ccefba31ebc6bb164f3b9821adf2c559.tar.gz";
      # Hash obtained using `nix-prefetch-url --unpack <url>`
      sha256 = "0zg6w45yq9aacff8s76a8jv6b1hvp988mb28r495rpbl5jdapna5";
    })
    { }
}:

with bad-pkgs;
mkShell {
  buildInputs = [
    (neovim.override {
      configure = {
        packages.myVimPackage = with vimPlugins; {
          start = [ ultisnips ];
          opt = [ ];
        };
      };
    })
  ];
}
```
You can change the `with bad-pkgs;` to `with good-pkgs;` where that problem is fixed.

**Comment by SirVer** [comment]
@KoviRobi good practice if you want to have bugs being looked at is for maintainers to make reproducing them as easy as possible. Using an esoterice third party tool for configuration is frankly not this. I will not try to repro this with `nixpkgs`, please provide me a repro case as asked in my usual request text:

Please provide a fully reproducible example, ideally using the approach explained in https://github.com/SirVer/ultisnips/blob/master/CONTRIBUTING.md#reproducing-bugs or if not possible with a minimal `vimrc` and snippets files and an exact sequence of key strokes that trigger the bug.

**Comment by KoviRobi** [comment]
First of all, you are right, I did say I will provide a docker image on request, then you did ask for a docker image and I didn't give you one, so here it is
https://hub.docker.com/r/kovirobi/nvim-ultisnips-ins-del-hang/tags with two tags, one for working and one for broken.

Before you jump in though, just to make it clear, the bug seems to originate from Pynvim, that's why I didn't provide the docker image when you asked, because the bug was fixed in pynvim (and that's why you couldn't reproduce it). I do tend to ramble though, apologies.
